pulsar/producer.go

```diff
@@ -138,6 +138,9 @@ type ProducerOptions struct {
 
 	// A chain of interceptors, These interceptors will be called at some points defined in ProducerInterceptor interface
 	Interceptors ProducerInterceptors
+
+	// MaxReconnectToBroker set the maximum retry number of reconnectToBroker. (default: no retry)
+	MaxReconnectToBroker int
```

Comment:
@jonyhy96 Can we not modify the original default behavior, and we will always retry by default. When we set the maximum number of retries, we will turn off the retry logic after reaching the specified threshold.

Because the current behavior will break the backward compatibility of the version

Reply:
Sure, neither do I want to break the current behavior. The original commit of this pr will fulfill our request, please check out the [comments](https://github.com/apache/pulsar-client-go/pull/360#discussion_r476231877) above. @shohi may have different opinion on this due to his(her) proposal. Please take a review.

Reply:
We can consider setting -1 as the default value, and then check whether MaxReconnectToBroker is set, and if so, use the user-defined value.

Reply:
At present, I see that you have added the option of MaxReconnectToBroker to the relevant test cases. This may not be very user-friendly. We need to ensure that the interface is as simple as possible.

Reply:
> We can consider setting -1 as the default value, and then check whether MaxReconnectToBroker is set, and if so, use the user-defined value.

I understand,but how can we determine whether MaxReconnectToBroker is set or not when it's default value is meaningful. e.g. user set this option to 0.

> At present, I see that you have added the option of MaxReconnectToBroker to the relevant test cases. This may not be very user-friendly. We need to ensure that the interface is as simple as possible.

I totally agree, how about we use the origin commit of this pr to implement this feature?

Reply:
ok, agree with you

Reply:
Should i reset this pr to original commit or keep it this way?

Reply:
@jonyhy96 Sorry for the reply later, i think we should reset this pull request to the original commit for backward compatibility.

Reply:
> @jonyhy96 Sorry for the reply later, i think we should reset this pull request to the original commit for backward compatibility.

Already reset this pr to original commit.
By the way, the go.mod file seems a little bit dirty,you can reproduce this by run comamnd `go mod tidy` locally.

```shell
➜  pulsar-client-go git:(feat-max-reconnect-to-broker) go mod tidy -v        
unused github.com/modern-go/concurrent
unused github.com/modern-go/reflect2
```

perhaps add script below to ci for check if go.mod has been commit correctly

```shell
go mod tidy
git diff --exit-code -- go.mod go.sum
```